Loosen type check on parameter act

martinschumann opened the pull request:
Hi Andreas!

Highly appreciating an heavily using ce-access we ran into this:

Access to an article in some cases is denied by mistake and redirects to error message caused by strict comparison on parameter `'act'`.  

Specifically:
- if `tl_content` contains a record with the same ID as the article to be listed with its content elements and 
- current user is not allowed to edit the specific type of this unconnected content element.

This happens since param `'act'` is now validated in a negated strict comparison and the following filtering block and final redirect exit will be applied when listing the content elements of an article (`'act' == NULL`, and therefore not a `string`). For this request `$dc->id` contains the ID of an article and must not be compared to IDs of table `tl_content`.

Since `Input::get` always returns either `string` or `NULL` I would propose to discard strict comparison later on as well.

Regards
Martin

Commit:
Loosen type check on parameter act

aschempp commented:
I don't like to remove strict comparison, that's bad practice. How about replace it with `(string) \Input::get('act') !== ''` instead?

Commit:
Remove denial of access

Ignore unset query string key "act" while staying strict in comparison to remove falsely denied access.

aschempp commented:
There is no need to typecast the non-empty value, as the comparison can never be incorrectly true. Also, `empty` on method return values is only availavle since PHP 5.5 (afair) but we're still supporting 5.3 here. One typecast for the empty string should be totally enoght 😊

Commit:
Remove denial of access in most simple way

Ignore unset or empty query string key "act" while staying strict in comparison

Commit:
Clean up

martinschumann commented:
Mhm. :-) Okay, `empty()` can't be called yet on functions in PHP < 5.5.

To exclude applying the filtering block for a query string key 'act' not being present, it's sufficient to evaluate `\Input::get('act')` directly.

Key not present in URL => NULL
Key present in URL without any value => empty string

Since $_GET contains strings only, typecasting is not necessary at all.

What about keeping it simple like this?

aschempp commented:
Why don't you just do a `(string) \Input::get('act') !== ''` ? 😂

Commit:
Remove denial of access

martinschumann commented:
That's a good question… :-) I'll ask myself later.

martinschumann commented:
You're right. My mistake and wrong assumption: thought typecasting NULL would emit E_NOTICE and wanted to suppress that. Sorry for confusion.

aschempp commented:
Fixed in 2fcfa40c416380c0089f77ce7fd6987887b9bd11
